### ISSUE TITLE
NextJS emulators:start --only functions,hosting bug

### ISSUE THREAD

**Issue by jasdevsidhu12** [firebase/firebase-tools#3055]
Hi There,
The following is my branch and I'm trying to create a simple firebase cloud "hosting, functions" for my NextJS application.
_Repo_
**https://github.com/jasdevsidhu12/test-next**
_Branch_ 
**firebase-issue**

For the whole of last week I kept trying and trying but I couldn't figure it out why `firebase emulators:start --only functions,hosting` **wouldn't work** but  `firebase deploy --only functions,hosting` **worked.**

**Result of  `firebase emulators:start --only functions,hosting`**
![image](https://user-images.githubusercontent.com/12964252/104815829-fa79ba00-580e-11eb-9dbe-13a7f993ea79.png)
firebase-debug.log(**No errors shown**)
![image](https://user-images.githubusercontent.com/12964252/104816073-a40d7b00-5810-11eb-8569-7b46410c230a.png)

**Result of `firebase deploy --only functions,hosting`**
![image](https://user-images.githubusercontent.com/12964252/104815854-3745b100-580f-11eb-86ed-f4e5d9904908.png)

To replicate bug

- clone the Repo

- `npm run serve`

**Comment by yuchenshi** [comment]
Would you mind trying `firebase emulators:start --only functions,hosting --debug` and post the longer debug log? There may be some interesting things in the fuller log.

**Comment by samtstern** [comment]
@jasdevsidhu12 in addition to the logs @yuchenshi requested can you show your `firebase.json`?  Generally when using a framework like React or Vue you don't use `localhost:5000` as your dev server because that will serve the files listed in your hosting public folder which is normally something like `app/dist` or another folder of production built files.  Instead you should use the local web server built in to Next.js

**Comment by jasdevsidhu12** [comment]
Hi @yuchenshi thanks for replying quick :)
Here's the log https://github.com/jasdevsidhu12/test-next/blob/firebase-issue/firebase-debug.log . I couldn't find anything interesting :( 

Hi @samtstern here's my firebase.json file https://github.com/jasdevsidhu12/test-next/blob/firebase-issue/firebase.json
about ```localhost:5000``` I think ```firebase emulators:start --only functions,hosting``` assigns it. I'm not sure whether I can change it. 
If I use **NextJS Dev server which normally host is ```localhost:3000``` it doesn't have any issues** but I would love to test the whole application before deploying it to firebase.

**Comment by samtstern** [comment]
@jasdevsidhu12 if you look at your `firebase.json` you'll see this:

```
{
  "hosting": {
    "public": "server/dist",
```

This means Firebase Hosting (in prod and when emulated) will serve the contents of `server/dist`. This is generally the output of `npm run build`. The hosting emulator does not proxy the contents of the Next.js dev server.

So while developing you should continue to use the NextJS dev server. If you want to test before deploying first build your website for production so it's in `server/dist` and then check out the hosting emulator.

There's some discussion of this issue here:
https://github.com/firebase/firebase-tools/issues/594#issuecomment-609674903

You *may* be able to point the Next.js dev server at `server/dist` in which case everything should just work. For now I am closing this as intended behavior, let me know if I missed something!

**Comment by jasdevsidhu12** [comment]
Hi @samtstern,
I'm using an **express-server** to serve/render those NextJS files and its in my repository the [server file.](https://github.com/jasdevsidhu12/test-next/blob/firebase-issue/server/server.ts). 

My **express-server** is using the firebase **cloud-function/hosting** name **helloWorld** [line 20] to **"run on"**. The cloud-function name is also stated in  my [firebase.json file](https://github.com/jasdevsidhu12/test-next/blob/firebase-issue/firebase.json) [line 7]. 

So when I run my **express-server/cloud-functions/hosting** locally  using  ```firebase emulators:start --only functions,hosting``` command it produces the following error **(image below)**.
![image](https://user-images.githubusercontent.com/12964252/105211687-78f29680-5b44-11eb-9f02-9285ecc87d32.png)


However, if I run my **express-server only without using cloud-functions/hosting it works without any issues(images below)** . 
(screenshot from my [server file.](https://github.com/jasdevsidhu12/test-next/blob/firebase-issue/server/server.ts))
![image](https://user-images.githubusercontent.com/12964252/105208871-16e46200-5b41-11eb-8aff-df144d38a245.png)
![image](https://user-images.githubusercontent.com/12964252/105208423-945ba280-5b40-11eb-89d6-2a6d0f94f6b7.png)

**Results (pure express-server)**
![image](https://user-images.githubusercontent.com/12964252/105206333-28783a80-5b3e-11eb-909b-30a49609f6a0.png)


Moreover the same **express-server/cloud-functions/hosting** is what I deployed to my firebase ```fir-nextjs-note-app``` application and it works over there (from my 1st comment).  So in a nutshell, the **express-server/cloud-functions/hosting** breaks(cannot render the html file) when I run ```firebase emulators:start --only functions,hosting``` locally. Can you help me out and is this issue still close?
 I appologise I should have given more background on this issue.

**Comment by samtstern** [comment]
Can you show the full request and response (url, headers, data, etc) from the `test` request you circled in the Chrome console?

Also it looks like you're trying to send `/test` to the `helloWorld` function but your `firebase.json` does not have the configuration for that:
https://firebase.google.com/docs/hosting/functions

You need to add something like this:

```json
"hosting": {
  // ...

  // Add the "rewrites" attribute within "hosting"
  "rewrites": [ {
    "source": "/test",
    "function": "helloWorld"
  } ]
}
```

**Comment by jasdevsidhu12** [comment]
so when I hit request ```/test``` the following are the headers. I'm not passing any query.
![image](https://user-images.githubusercontent.com/12964252/105220504-b9a3dd00-5b4f-11eb-8098-aa5a9d477dff.png)

and in cmd log 
![image](https://user-images.githubusercontent.com/12964252/105220593-da6c3280-5b4f-11eb-819d-c31b68a37940.png)
Still cant find anything unusual

Also on the URL rewrites I want all request to hit the helloWorld express-server/cloud-functions, hence ```"source": "**"```

**Comment by bkendall** [comment]
Hi folks. I was able to replicate this issue and I think the root cause is similar to #3052.

When the hosting emulator is streaming a response, there's a case where the caller may be expecting compressed content and the proxy is _decompressing_ the content and streaming that back as a response. This causes issues where it can't decompress the content (because it's not compressed) and the content looks truncated (because it's coming back with a content-length of the compressed response rather than the raw response.

I'm working on a PR that should address this issue (I was able to verify that the fix worked in this case too, if I remember yesterday correctly). I'm making sure to add some tests to make sure it's behaving as expected, too.

Thanks for your patience.
